pkg/kv/kvserver/replica_learner_test.go

```diff
@@ -1711,44 +1712,42 @@ func getExpectedSnapshotSizeBytes(
 	}
 	defer snap.Close()
 
-	totalBytes := int64(0)
-	var b storage.Batch
-	defer func() {
-		b.Close()
-	}()
-	b = originStore.Engine().NewUnindexedBatch(true)
-	iter := snap.Iter
-	var ok bool
-	for ok, err = iter.SeekStart(); ok && err == nil; ok, err = iter.Next() {
-		hasPoint, hasRange := iter.HasPointAndRange()
-		if hasPoint {
-			var unsafeKey storage.EngineKey
-			if unsafeKey, err = iter.UnsafeKey(); err != nil {
-				return 0, err
-			}
-			if err := b.PutEngineKey(unsafeKey, iter.UnsafeValue()); err != nil {
-				return 0, err
-			}
-		}
-		if hasRange {
-			bounds, err := iter.RangeBounds()
-			if err != nil {
-				return 0, err
-			}
-			for _, rkv := range iter.RangeKeys() {
-				err := b.PutEngineRangeKey(bounds.Key, bounds.EndKey, rkv.Version, rkv.Value)
-				if err != nil {
-					return 0, err
+	b := originStore.Engine().NewUnindexedBatch(true)
+	defer b.Close()
+
+	err = rditer.IterateReplicaKeySpans(snap.State.Desc, snap.EngineSnap, true, /* replicatedOnly */
+		func(iter storage.EngineIterator, _ roachpb.Span, keyType storage.IterKeyType) error {
+			var err error
+			for ok := true; ok && err == nil; ok, err = iter.NextEngineKey() {
```

Comment:
In store_snapshot.go we do that for performance, to avoid additional branching in the hot path. We don't care about performance in test code, and this is slightly more compact, but I don't have a strong opinion either way.

>Overall it seems a bit unfortunate that we have to cram two functions into one as they usually have two separate parts to handle key and ranges.

I figure if the logic is non-trivial then the caller can dispatch to separate functions themselves.